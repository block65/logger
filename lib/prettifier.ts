import {
  bgRed,
  bgYellow,
  black,
  blue,
  bold,
  gray,
  green,
  red,
  whiteBright,
} from 'colorette';
import { isatty } from 'tty';
import util from 'util';
import type { LogDescriptor } from './types.js';

function formatLevel(level: number): string {
  switch (level) {
    case 60: //fatal
      return whiteBright(bgRed(bold('FATAL')));
    case 50: //error
      return red('ERROR');
    case 40: // warn
      return bgYellow(black(bold('WARN')));
    case 30: //info
      return blue('INFO');
    case 20: //debug
      return green('DEBUG');
    case 10: //trace
      return gray('TRACE');
    default:
      return gray('SILENT');
  }
}

export function createPrettifier(/*options?: unknown*/) {
  return (log: LogDescriptor): string => {
    const { level, msg = '', time, name, hostname, pid, ...rest } = log;

    if (msg?.startsWith('pino.final with prettyPrint')) {
      return '';
    }

    const formattedName = name ? `(${name})` : '';
    const formattedMsg = msg && ' ' + whiteBright(msg);

    const formattedRest =
      Object.keys(rest).length > 0
        ? ' ' +
          util.inspect(rest, {
            colors:
              'FORCE_COLOR' in process.env ||
              (isatty(1) &&
                process.env.TERM !== 'dumb' &&
                !('NO_COLOR' in process.env)),
            compact: true,
            sorted: true,
          })
        : '';

    return `${gray(new Date(time).toJSON())} ${formatLevel(level).padEnd(
      6,
      ' ',
    )}${formattedName} ${formattedMsg}${formattedRest}\n`;
  };
}
