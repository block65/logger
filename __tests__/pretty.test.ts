import { describe, test, expect, jest, beforeEach } from '@jest/globals';
import { createPrettifier } from '../lib/prettifier.js';
process.env.NO_COLOR = 'true';

describe('Prettifier', () => {
  const OLD_ENV = process.env;
  beforeEach(() => {
    jest.resetModules();
    process.env = { ...OLD_ENV };
  });

  test('Color', async () => {
    process.env.NO_COLOR = 'true';

    const prettifier = createPrettifier();

    await expect(
      prettifier({
        level: 20,
        pid: 1,
        time: 1234567890,
        hostname: 'test',
        msg: 'kek',
      }),
    ).toMatchSnapshot();

    await expect(
      prettifier({
        level: 20,
        pid: 1,
        time: 1234567890,
        hostname: 'test',
        msg: 'kek',
        random: {
          bool: true,
          str: 'hello',
          number: 12,
          arr: [1, 2, 3],
        },
      }),
    ).toMatchSnapshot();
  });

  test('With colour', async () => {
    process.env.FORCE_COLOR = 'true';

    const prettifier = createPrettifier();

    await expect(
      prettifier({
        level: 20,
        pid: 1,
        time: 1234567890,
        hostname: 'test',
        msg: 'kek',
      }),
    ).toMatchSnapshot();

    await expect(
      prettifier({
        level: 20,
        pid: 1,
        time: 1234567890,
        hostname: 'test',
        msg: 'kek',
        random: {
          bool: true,
          str: 'hello',
          number: 12,
          arr: [1, 2, 3],
        },
      }),
    ).toMatchSnapshot();
  });

  // test('via import', async () => {
  //   // NOTE: All we can do here is make sure this doesn't throw.
  //   // It's not possible to test via Jest because we can't provide both a
  //   // transport AND a stream and because transports are just modules, all we
  //   // can do is mock it, which defeats the purpose of testing it
  //
  //   const { createLogger } = await import('../dist/index.js');
  //
  //   const logger = createLogger({
  //     pretty: true,
  //   });
  //
  //   logger.trace({
  //     test: 123,
  //   });
  // });
});
