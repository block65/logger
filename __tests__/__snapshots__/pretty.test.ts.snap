// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prettifier Color 1`] = `
"1970-01-15T06:56:07.890Z DEBUG   kek
"
`;

exports[`Prettifier Color 2`] = `
"1970-01-15T06:56:07.890Z DEBUG   kek { random: { arr: [ 1, 2, 3 ], bool: true, number: 12, str: 'hello' } }
"
`;

exports[`Prettifier With colour 1`] = `
"1970-01-15T06:56:07.890Z DEBUG   kek
"
`;

exports[`Prettifier With colour 2`] = `
"1970-01-15T06:56:07.890Z DEBUG   kek { random: { arr: [ [33m1[39m, [33m2[39m, [33m3[39m ], bool: [33mtrue[39m, number: [33m12[39m, str: [32m'hello'[39m } }
"
`;
